packages/liveblocks-client/src/LiveMap.ts

```diff
@@ -13,22 +13,23 @@ import {
   CrdtType,
   SerializedCrdt,
 } from "./live";
-import { StorageUpdate } from "./types";
+import { LiveMapUpdates } from "./types";
+import { Lson } from "./lson";
 
 /**
  * The LiveMap class is similar to a JavaScript Map that is synchronized on all clients.
  * Keys should be a string, and values should be serializable to JSON.
  * If multiple clients update the same property simultaneously, the last modification received by the Liveblocks servers is the winner.
  */
-export class LiveMap<TKey extends string, TValue> extends AbstractCrdt {
-  private _map: Map<TKey, AbstractCrdt>;
+export class LiveMap<TValue extends Lson = Lson> extends AbstractCrdt {
```

Comment:
I don't have a strong opinion about it but I feel like it's more explicit when you have to declare the type of the key and value. Like it's clear that it's a map.

Reply:
Yeah. The symmetry with `Map<K, V>` is definitely nice, and it initially feels like a weird sacrifice. However, with `Map<K, V>` you can use anything as keys (object instances, strings, numbers, etc).

But with a `LiveMap` you cannot, and this new type honestly reflects that. Since `LiveMap` keys can only really be `string`s (implied currently anyway by the `TKey extends string` constraint), as a user you must already always explicitly type/provide `string` as the first type param. If you have a case for limiting `TKey` to only certain string literals (think `LiveMap<'foo' | 'bar', X>`), I think a `LiveObject` would be a better fit by design.

Another goal here is to help users better pick which one to use, because right now it's a little unclear for what cases you would pick one over the other. The documentation doesn't help to make that choice, because both descriptions are the same:

![Screen Shot 2022-04-15 at 19 47 51@2x](https://user-images.githubusercontent.com/83844/163603894-431ea673-5763-4bda-a450-8a6674a4f86a.png)

Hopefully this can help make it a bit clearer. I will definitely also update the docs to better help illustrate the differences and for what use cases to pick which data structure. Hope this helps to make this feel a little less weird.

Reply:
Yeah I agree that our API/documentation doesn't correctly explain when to use LiveMap vs LiveObject.

I'm still not sure about that breaking change. Do you think we could ask the community on Discord what they think about it?

Reply:
I remember wanting to remove the "key" param at some point 😄 Here is the TODO I put a long time ago:

https://github.com/liveblocks/liveblocks/commit/608382dae3de7f0db0a762dca073337d762f3ba4#diff-273c9393df39cefe5414e6fc728a8da73aee11cce8507a7288ac8d22ea89a63dR523

For the backstory, the first implementation of the storage didn't include `LiveMap`. It was only `LiveObject` (called `LiveRecord` at the time)  and `LiveList`. I had to help a few early users that wanted `LiveObject` with dynamic keys (not fully known at compile time). I quickly realized `LiveObject<{ [key: string]: string }>` or `LiveObject<Record<string, string>>` were not intuitive for a significant percentage of typescript developers. Hence I created `LiveMap` for dynamic keys.

So why I finally decided to keep the first argument?

- Symmetry with `Map` in typescript, as mentioned by @ofoucherot. We're only missing the `clear` method (could be implemented but might require a special server op) and iterability-in-order (will probably never be implemented). 
- We might want to support key as `number` in the future. It's been a while since I tested this but I think it should be relatively easy to implement and wouldn't introduce a breaking change.
- We support template literal types. For example, it could be useful for a spreadsheet.

```
type SpreadsheetKey = `${Uppercase<string>}:${number}`
const livemap = new LiveMap<SpreadsheetKey, string>();

livemap.set("C:2", "Hey");
livemap.set("invalidKey", "Hey");
```

https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html


- Subjectively, having 2 generic arguments for the `LiveMap` and 1 for the `LiveObject` make `LiveMap.get` vs `LiveObject.get` or `LiveMap.set` vs `LiveObject.set` signatures intuitive to me without looking at it.
- Related to the first point, all the Map implementations I could find in different languages take 2 generic arguments. (Sometimes called HashMap or Dictionnary).

I'm a bit reluctant to introduce this breaking change because it might impact a significant percentage of our users, but happy to brainstorm the pros and cons if it makes the API more intuitive for future users. :)

Reply:
I think I went in a bit too aggressively there, and I will undo that change by reintroducing the `TKey extends string` type param there 👍 The arguments for retaining symmetry, enabling string literal types and future support for numeric keys make total sense, and I definitely did not intend to close those roads off!

Reply:
Restored the key type in e45c468c7a107308d7ebab5ea567089351538581 — thx for the feedback 👍 !